File: src/services/urlExtractor.ts
import axios from 'axios';
import * as cheerio from 'cheerio';

export async function extractFromUrl(url: string): Promise<string> {
  if (url.includes('twitter.com') || url.includes('x.com')) {
    return extractTwitter(url);
  } else if (url.includes('reddit.com')) {
    return extractReddit(url);
  } else {
    return extractGeneric(url);
  }
}

async function extractTwitter(url: string): Promise<string> {
  try {
    // Try fxtwitter.com first as it's optimized for embeds/meta tags
    const fxUrl = url.replace(/(twitter\.com|x\.com)/, 'fxtwitter.com');
    const response = await axios.get(fxUrl, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
      }
    });
    const $ = cheerio.load(response.data);
    
    // fxtwitter often puts the text in og:description or a specific tag
    const description = $('meta[property="og:description"]').attr('content') || 
                        $('meta[name="description"]').attr('content');
    
    if (description) {
      return description.replace(/^“/, '').replace(/”$/, '').trim();
    }
    
    return "Could not extract tweet content. Twitter's protection prevents simple scraping. Please paste the text directly.";
  } catch (error: any) {
    return `Twitter Extraction Error: ${error.message}`;
  }
}

async function extractReddit(url: string): Promise<string> {
  try {
    // Reddit has a simple .json endpoint for public posts
    const jsonUrl = url.split('?')[0].replace(/\/$/, '') + '.json';
    const response = await axios.get(jsonUrl, {
      headers: {
        'User-Agent': 'JP-Norm-Bot/1.0'
      }
    });
    
    const postData = response.data[0].data.children[0].data;
    const title = postData.title || '';
    const body = postData.selftext || '';
    
    return `${title}\n\n${body}`.trim();
  } catch (error: any) {
    return `Reddit Extraction Error: ${error.message}`;
  }
}

async function extractGeneric(url: string): Promise<string> {
  try {
    const response = await axios.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0'
      }
    });
    const $ = cheerio.load(response.data);
    
    // Remove scripts, styles, nav, footer
    $('script, style, nav, footer, header').remove();
    
    const text = $('body').text();
    return text.replace(/\s+/g, ' ').trim().substring(0, 2000);
  } catch (error: any) {
    return `Generic Extraction Error: ${error.message}`;
  }
}

export function cleanExtractedText(text: string): string {
  // Remove URLs
  let cleaned = text.replace(/https?:\/\/\S+/g, '');
  // Remove extra whitespace
  cleaned = cleaned.replace(/\s+/g, ' ').trim();
  return cleaned;
}

File: url_extractor.py
import re
import requests
from bs4 import BeautifulSoup
import json

# Note: snscrape is often unstable due to Twitter changes. 
# In a real production environment, official APIs or robust scrapers are used.
# For this project, we implement a flexible extraction logic.

def extract_from_url(url: str):
    """
    Main entry point for URL extraction.
    Detects platform and routes to specific extractor.
    """
    if "twitter.com" in url or "x.com" in url:
        return extract_twitter(url)
    elif "reddit.com" in url:
        return extract_reddit(url)
    else:
        return extract_generic(url)

def extract_twitter(url: str):
    # Mock/Simplified extraction for Twitter if snscrape fails
    # Real snscrape implementation would be:
    # import snscrape.modules.twitter as sntwitter
    # tweet = next(sntwitter.TwitterTweetScraper(url).get_items())
    # return tweet.content
    
    try:
        response = requests.get(url, headers={"User-Agent": "Mozilla/5.0"})
        soup = BeautifulSoup(response.text, "html.parser")
        # Twitter often hides content from simple scrapers, but we try to find meta tags
        meta_desc = soup.find("meta", property="og:description")
        if meta_desc:
            return meta_desc["content"]
        return "Could not extract tweet content. Twitter requires authentication or advanced scraping."
    except Exception as e:
        return f"Twitter Extraction Error: {str(e)}"

def extract_reddit(url: str):
    # Simplified Reddit extraction using .json endpoint (no API key needed for public posts)
    try:
        if not url.endswith(".json"):
            json_url = url.split("?")[0].rstrip("/") + ".json"
        else:
            json_url = url
            
        response = requests.get(json_url, headers={"User-Agent": "JP-Norm-Bot/1.0"})
        data = response.json()
        
        post_data = data[0]["data"]["children"][0]["data"]
        title = post_data.get("title", "")
        body = post_data.get("selftext", "")
        
        return f"{title}\n\n{body}".strip()
    except Exception as e:
        return f"Reddit Extraction Error: {str(e)}"

def extract_generic(url: str):
    try:
        response = requests.get(url, headers={"User-Agent": "Mozilla/5.0"})
        soup = BeautifulSoup(response.text, "html.parser")
        
        # Remove script and style elements
        for script in soup(["script", "style"]):
            script.decompose()
            
        text = soup.get_text()
        # Clean up whitespace
        lines = (line.strip() for line in text.splitlines())
        chunks = (phrase.strip() for line in lines for phrase in line.split("  "))
        text = '\n'.join(chunk for chunk in chunks if chunk)
        
        return text[:2000] # Limit to first 2000 chars
    except Exception as e:
        return f"Generic Extraction Error: {str(e)}"

def clean_text(text: str):
    # Remove URLs
    text = re.sub(r'http\S+', '', text)
    # Remove extra whitespace
    text = re.sub(r'\s+', ' ', text).strip()
    return text
